### ISSUE TITLE
proxy jump failure in mina sshd 2.13.2

### ISSUE THREAD

**Issue by lawmin2021** [apache/mina-sshd#609]
### Version

2.13.2

### Bug description

the mina sshd 2.13.2 has problem to parse follow code:

private ClientSession getSessionUsingJumpHost(
    SshClient sshClient, CommandObject commandObject)
    throws OcmdException, IOException
  {
    String hostName = commandObject.getNode();
    String user = commandObject.getUser();
    int port = Constants.getSSHPort();

    // configure the bastion entry
    HostConfigEntry bastion = new HostConfigEntry();
    bastion.setHost(Constants.getTunnelHost());
    bastion.setHostName(Constants.getTunnelHost());
    bastion.setUsername(Constants.getTunnelHostUser());
    bastion.setPort(port);
    bastion.setIdentities(
      Collections.singleton(Constants.getTunnelHostUserKey()));

    // configure the generate target entry
    HostConfigEntry targetHost = new HostConfigEntry();
    targetHost.setHost("*"); // for all include hostname/ip
    targetHost.setProxyJump(Constants.getTunnelHost());

    HostConfigEntryResolver resolver =
      HostConfigEntry.toHostConfigEntryResolver(
        GenericUtils.asList(bastion, targetHost));

    sshClient.setHostConfigEntryResolver(resolver);

    // enable the public auth
    sshClient.setUserAuthFactories( Collections.unmodifiableList(
      Arrays.asList(
        UserAuthPublicKeyFactory.INSTANCE,
        UserAuthPasswordFactory.INSTANCE,
        UserAuthKeyboardInteractiveFactory.INSTANCE
      )));

    // now try to connect using proxy
    try
    {
      ClientSession session = sshClient.connect(user, hostName, port)
        .verify(sessionConnectTimeout)
        .getSession();
      return session;
    }
    catch (Exception ex)
    {
      throw  ex;
    }
  }

above code work in mina sshd 2.12.1.

### Actual behavior

The Mina SSHD code in SshClient.parseProxyJumps(HostConfigEntry entry, AttributeRepository context) method has an issue with identifying the variable "additionalHops" , and it causes the list object "hops" continuously grow to the maximum limit of 10, and then throw  the IlllegalArgumentException.
![issue1](https://github.com/user-attachments/assets/92ca3140-61ba-4cdf-9107-474eee75dfb7)
![issue2](https://github.com/user-attachments/assets/2ae596a4-a0f5-4e2b-9326-788eabef3f3b)

please check the attached debug images

### Expected behavior

I think in SshClient.parseProxyJumps() need to check if the return values of "additionalHops" is  already in the "hops", it need to stop and return without throw exception

**Comment by lawmin2021** [comment]
I made the following change to pass my case, hope it work for [GH-318](https://github.com/apache/mina-sshd/issues/318) Handle cascaded proxy jumps

protected List<HostConfigEntry> parseProxyJumps(HostConfigEntry entry, AttributeRepository context)
            throws IOException {
        List<HostConfigEntry> hops;
        try {
            hops = parseProxyJumps(entry.getProxyJump(), context);
            if (GenericUtils.isEmpty(hops)) {
                return hops;
            }
            // If the last entry itself has proxy jumps, we need to append them. Guard against possible proxy jump
            // loops by imposing an upper limit on the total number of jumps.
            for (;;) {
                HostConfigEntry last = hops.get(hops.size() - 1);
                try {
                    List<HostConfigEntry> additionalHops = parseProxyJumps(last.getProxyJump(), context);
                    if (additionalHops.isEmpty()) {
                        break;
                    } else {
                        boolean isFound = false;
                        for (HostConfigEntry checkAddHop : additionalHops) {
                            for (HostConfigEntry checkHop : hops) {
                                String fullHopStr = checkHop.toString();
                                String fullAddHopStr = checkAddHop.toString();
                                if (fullHopStr.equals(fullAddHopStr)) {
                                    isFound = true;
                                    break;
                                }
                            }
                            if (!isFound) {
                                hops.add(checkAddHop);
                            }

                        }

                        if (isFound) {
                            return hops;
                        }

                        if (hops.size() > CoreModuleProperties.MAX_PROXY_JUMPS.getRequired(this)) {
                            throw new IllegalArgumentException("Too many proxy jumps for host " + entry.getHost());
                        }
                    }
                } catch (IOException | RuntimeException e) {
                    throw new IllegalArgumentException("Problem parsing proxyJump from host config " + last.getHost(), e);
                }
            }
        } catch (IOException | RuntimeException e) {
            throw new IllegalArgumentException("Problem parsing proxyJump from host config " + entry.getHost(), e);
        }
        return hops;
    }

**Comment by tomaswolf** [comment]
The problem is your "*" config. You should exclude the bastion itself there. Would your config as shown work with OpenSSH?
```
Host bastion.example.com
Hostname bastion.example.com
Port 22
Username someone
Identity bastion_id

Host *
ProxyJump bastion.example.com
```
I guess this won't work well with OpenSSH either. It should work if the last entry is changed to
```
Host *, !bastion.example.com
ProxyJump bastion.example.com
```

**Comment by lawmin2021** [comment]
yes, SSH does not support that.
But from this page:  https://man.openbsd.org/ssh_config
"A single ‘*’ as a pattern can be used to provide global defaults for all hosts."

It really depends on the implementation. The Mina SSHD version 2.12.1 supports the '*' wildcard

I believe it would be beneficial to support the '*' wildcard programmatically. For instance, in my environment, there’s no specific pattern to identify the hostname or IP address. If '*' is supported, we wouldn't need to worry about the details of the remote hosts.

So just because the Linux SSH command doesn't support it doesn't mean that the Java library shouldn't

**Comment by lawmin2021** [comment]
The following code changes works, based on your example

 HostConfigEntry targetHost = new HostConfigEntry();
    targetHost.setHost("*,!" +Constants.getTunnelHost()); // for all include hostname/ip
    targetHost.setProxyJump(Constants.getTunnelHost());

**Comment by tomaswolf** [comment]
We do support the "*" wildcard.

The configuration you were using was broken. If you instruct SSH to use a proxy jump through the bastion for _all hosts, including the bastion_, then of course you get an infinite recursion. That is exactly why we limit the length of the proxy chain.

Apache MINA sshd works as designed; there is nothing to fix in that respect in the code. Your configuration was broken. (That it "worked" in before 2.13 was actually a bug that got fixed in 2.13.)
